### PR TITLE
Initial creation of filesystem scanner for DMLib

### PR DESCRIPTION
**Pull request by rnpmsft**
# All SDK Contribution checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.
- [x] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [x] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [ ] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/master/eng/mgmt/mgmtmetadata).
- [ ] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running `generate.ps1`. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version.
	
	**Note: We have recently updated the PSH module called by `generate.ps1` to emit additional data. This would help reduce/eliminate the Code Verification check error. Please run following command**:

	    `dotnet msbuild eng/mgmt.proj /t:Util /p:UtilityName=InstallPsModules`

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

### CODE REVIEW

File: sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs
Comment (amnguye) [inline sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs:42]: We should throw the exceptions that get thrown by `GetFullPath` instead of breaking and swallowing the exception.

Also we should add a TODO to log this error as well.

File: sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs
Comment (amnguye) [inline sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs:68]: If we lack the permissions we should somehow log this or maybe throw an exception. Maybe let's add a TODO here to add a logging.

File: sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs
Comment (amnguye) [inline sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs:84]: Same here, add a TODO for logging, it would be good for the user to know why we didn't upload certain files.

File: sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs
Comment (amnguye) [inline sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs:78]: Is the reason why you're doing this is to keep the stream open? Maybe we should do this check later when we actually have to open the file because Murphy's Law (people can go and change the access control on the file before we go to upload even though we checked here).

Comment (rnpmsft) [reply sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs:78]: I thought that the stream automatically closes once that using block is exited, is that wrong? I did it because I was using File.Exists(path) before since it claimed to be able to discern read permissions in the documentation, but that didn't hold up when I was testing (it just told us whether or not the file literally existed) and I needed another approach.

So, it seemed like one option was looking through ACL for the file, but then I saw a file could deny reading ACL while allowing reading contents. Instead, I opted to just attempt to open a stream with read permissions, as it should use that check to know whether or not to add it to the enumeration for now, and then close the stream.

If it isn't possible to read later, I was assuming that could just be handled with log + skip per user options, as I wasn't sure if there was any better alternative to dealing with file changes from scan to operation time.

Comment (amnguye) [reply sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs:78]: I should have clarified my question. In the body of this stream you have it jump to `readSuccess` which is fine, but is there a reason why you did this over just returning the file in the body of this, or even just checking if it just opens and then closing it right after.
e.g. something like this
```
try
{
    File.Open(this.DataSourceFileName, FileMode.Open, FileAccess.Read).Dispose();
    return true;
}
```
Also is there a reason why we picked `Open` over `OpenRead`?

If it's not possible to read later, then this check is unnecessary right? We only care if we can access the file when we actually want to read it.

File: sdk/storage/Azure.Storage.Common.DataMovement/tests/FilesystemScannerTests.cs
Comment (amnguye) [inline sdk/storage/Azure.Storage.Common.DataMovement/tests/FilesystemScannerTests.cs:22]: Let's use something like this https://docs.microsoft.com/en-us/dotnet/api/system.io.path.gettemppath?view=net-5.0&tabs=windows

Also this to create files
https://docs.microsoft.com/en-us/dotnet/api/system.io.path.gettempfilename?view=net-5.0

Comment (amnguye) [reply sdk/storage/Azure.Storage.Common.DataMovement/tests/FilesystemScannerTests.cs:22]: Same for the other tests

File: sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs
Comment (rnpmsft) [reply sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs:78]: Nope, no reason for either of those. I'm changing the scanner to function as you wrote; I think I was overthinking something about the try-catch and yield return going together.

I suppose so; I was thinking it would save time for the operations later, but really it wouldn't be that intensive at all. So, the only files the scanner really wouldn't be able to return are those inside of a subfolder that we can't read, since we wouldn't be able to get inside to enumerate. However, if a file exists somewhere we are able to at least see that it exists, we want to put it in the list and leave it to the method caller to decide whether reads will or won't happen. Does that sound like the desired functionality? I'll drop the read checks here if that's the case.

File: sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs
Comment (rnpmsft) [reply sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs:42]: Any specific logging technique to use, or should I just leave it to Console.WriteLine()?

File: sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs
Comment (rnpmsft) [reply sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs:80]: TODO added for logging; I set it to throw if we get an exception trying to enumerate the parent folder, but skip and continue for subfolders.

I'm thinking that if a user wants to enumerate a folder, we'll first either be able to read the given path or not, and if we can, we'll either be able to read all or some of the subfolders inside (provided they exist). A user might want to only enumerate if every file is possible to enumerate, or just enumerate whatever is possible at the path supplied. I set the scanner to work for the latter case for now, but maybe we can make this an option to be set by the user (single argument or in options object)?

(Should I set it to the former case for now instead, or is this fine?)

File: sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs
Comment (rnpmsft) [reply sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs]: I took the file checking code out of the scanner for now, to avoid any mismatches between scan time and operation time as you mentioned. The functionality now is to simply return all files we have permission to see exists, and let the caller handle the question of whether or not we can work on the file.

File: sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs
Comment (rnpmsft) [reply sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs]: File check removed, see below.

File: sdk/storage/Azure.Storage.Common.DataMovement/tests/FilesystemScannerTests.cs
Comment (rnpmsft) [reply sdk/storage/Azure.Storage.Common.DataMovement/tests/FilesystemScannerTests.cs]: Rewrote tests to use this, all seem to pass on Windows.

TODO: Testing on Linux/macOS

I'm thinking the build errors from the pipelines at the moment might be coming from the tests. The Mono.Posix.NETStandard package I added is causing the following warning to show:
- MSB3245:	Could not resolve this reference. Could not locate the assembly "Mono.Posix". Check to make sure the assembly exists on disk. If this reference is required by your code, you may get compilation errors.
  - Azure.Storage.Common.DataMovement.Tests
  - Azure.Storage.Blobs.DataMovement.Tests

Windows builds fine, but since the package is only relevant for Unix/Posix platforms, it could be causing the build failures due to missing assembly. I need to set up the VM and check on this + run tests.

File: sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs
Comment (amnguye) [reply sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs]: Definitely still add the TODO, but also we should throw any exceptions caught.

Comment (rnpmsft) [reply sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs]: Added the TODO, and set it to throw anything that comes up here.

File: sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs
Comment (amnguye) [reply sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs:80]: Yeah that's a good idea, we will just follow what the user wants to do in the case. This is fine.

**Review by amnguye — APPROVED** [review]

File: eng/Packages.Data.props
Comment (kasobol-msft) [inline eng/Packages.Data.props:205]: We should check if we can introduce these dependencies first.
My guess is that we can't as they appear to be 3rd party.

File: sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs:21]: https://objcsharp.wordpress.com/2013/07/08/why-static-code-is-bad/

File: eng/Packages.Data.props
Comment (kasobol-msft) [reply eng/Packages.Data.props:205]: If these are only for testing we can add them in test project file directly.

File: sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs
Comment (rnpmsft) [reply sdk/storage/Azure.Storage.Common.DataMovement/src/FilesystemScanner.cs:21]: #21715 I refactored the scanner to make it non-static and shift a couple things around, let me know if there's any issues with this implementation.

File: eng/Packages.Data.props
Comment (rnpmsft) [reply eng/Packages.Data.props:205]: #21715 I took the two dependencies I introduced for testing out of this file and used VersionOverride in the .csproj. The error spooked me and override sounded scary.

I can't tell if the Mono.Posix is/isn't 3rd party, since Microsoft is listed as an owner alongside some other users. There was another package called Mono.Posix.NETStandard with only Microsoft as an owner, but for some reason, that package was searching for the assembly "Mono.Posix". I couldn't think of anything to get that one to properly find the assembly without potentially causing other issues.

Do you think I should drop the dependency and maybe just execute bash + call chmod manually? I fiddled with manually loading the c library for chmod, but I couldn't do it with .NET Core 2.1 as a target.

**Pull request merged**
